### ISSUE TITLE
Constrained generic type parameters accept invalid updates

### ISSUE THREAD

**Issue by sliminality** [microsoft/TypeScript#60623]
### 🔎 Search Terms

`generic type variables unsafe modifications`, `generic type variables mutation`, `polymorphic function return type`, `generic type variable invalid assignment`, `polymorphic type variable unsound assignment`

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about [Generics](https://github.com/Microsoft/TypeScript/wiki/FAQ#generics)

### ⏯ Playground Link

- [Example using object spread](https://www.typescriptlang.org/play/?ts=5.7.2#code/PTAEBUE8AcFMGUDGAnAltALqARgVwOYBcoA4rAHaxqKgYyygBuAhms9gDawDOocyAW1RZc5bswBmDAQHsAJqgmpEzDKhliAdAChtIUACojxk6bPmLlq0b1gAogA9mA6F2IBVbqnL5QM7ABWsIhY3NDIsMxytrakqIwUoMxMrKjsXKAABg6ZoADuABYy3Ax0cKCovBgFDPgUVMq09ClsnAyZ4JkANLF5DCrkoGERURXkLByocqC40HKqPGMYMlk5+QWoGeE8VIzevsJVzR2ZOvrWF5dX5rplDACSAsx1oAC8oADe2qA-TXDEACJUE86gCer8hshEMRuBg0D4egBfW7NADCNUQAGtsDIHG9Pt9fndAYgMdjcWDCT9ScFMbA5MQcTIuMxyEjdBJRCF1INROIpAAecCgWAODAUOS8R7PBgAH1A6NpOIcAD4ABQOYjgACUWoJEIiGFwyEGH1AmgtDi6kOhoAAjKBEaB9KiOP5sFReBEAI64VARaZKWAcOQAQm0yO0iA0sIqIIY7z5klgarNxNAQPjYJtgIKrGZck0+EUAMd2t0wJlQ1UlSDUvjzrA0GY3BK3G0lbqmm4UOram4daGcP2jb4LbbUZjzNgmjd+DVnZnPcQmlJrAAghg1QAGbXa0eORCwTA84iL7tQ1d55Cbiq8cgyLDJTnkbkaIA) 
- [Example using `delete`](https://www.typescriptlang.org/play/?ts=5.7.2#code/PTAEBUE8AcFMGUDGAnAltALqARgVwOYBcoA4rAHaxqKgYyygBuAhms9gDawDOocyAW1RZc5bswBmDAQHsAJqgmpEzDKhliAdAChtIUACojxk6bPmLlq0b1gAogA9mA6F2IBVbqnL5QAAzlYLgxYP1tbUlRGClBmJlZUdi5-Bz9QAHcACxluBjo4UFReDEyGfAoqZVp6eLZOBj9wPwAaCPSGFXI+KgkZQULyFg5UOVBA4PUxUA0UtKzUZOhkHipGb19hYprGvx19awPDo-NdfIYAQVwSvtAAXlBtUCfQAB9QAG9quGIAIgFYATYKg-ZqgcjOWDEbgYNA+UG4XLIACScihMPWoAAvo9nm9PmdfvhcDwMCCwRC0bD8PDESiAPyUjHY7QSUSINQzcawEKeKgogA84FAsAcIXIcl4l2uyAAfAAKBzEcAASiVHxxTy5IVADk0CL5o30ACVYABHXCoZajXrIUD-QFUbgAQg1oGWGFwyC6Dm0zMQGmhdoBQNt9y1sF5yLkcveroJoD+weBrWe5P+v3gwwEINd+qjv0QAGY5ABOAAcEkLACYALRV9jYGsAFmYAFZWzXmE2SwB2GtybBlnutwISCSt2CwHOY5W6e0h0DiNTcJQ8UBS7K2-TQZjcXLcbTzqh62mjJdFVe8aFU0Db3f77T+sQyLiaDgyfByo-IE8GzSITJWHODA5QABmVZVb3sBxEFgTBJmIb9fyjQpeFEUdvFgOQgA)

### 💻 Code

Given a variable `x` whose type is generic parameter `T`, we can **spread invalid updates** alongside `x`, and the result is still assignable to `T`:

```ts
type Image = {
    type: "image",
    src: string,
}

type Checkbox = {
    type: "checkbox",
    checked: boolean,
}

function unsafe<T extends Image | Checkbox>(x: T): T {
    return { ...x, src: 1 } // Clobbers required field!
}

const image = unsafe({ type: "image", src: "harold.gif" })

image satisfies Image // passes
image.src satisfies string // passes
console.log(image.src.charAt(0)) // Exception: image.src.charAt is not a function
```

Likewise, we can **delete required fields** in `T`:

```ts
type Author = 
    | { type: "member", name: string, userId: string }
    | { type: "guest", name: string, userId?: string }

function deleteUserId<T extends Author>(x: T): T {
    delete x.userId // Required for members!
    return x
}

const member = deleteUserId({
    type: "member",
    name: "Slim",
    userId: "c3d98f32-2abb-4a55-a497-db875deff5ee",
})

member satisfies Author // passes
member.userId satisfies string // passes
console.log(member.userId.charAt(0)) // Exception: member.userId is undefined
```

### 🙁 Actual behavior

TypeScript does not error when clobbering or deleting a required field in a generically-typed value, and continues to infer the type of the original parameter, which is now invalid.

### 🙂 Expected behavior

I'd expect any of the following:
1. **Simpler, most conservative**: TypeScript prevents me from performing an update if it would invalidate _any_ possible `T`.
    - Attempting to **overwrite** a field required in any possible `T` with an invalid type produces an error like `Type 'number' is not assignable to type 'string'. (2322)`
    - Attempting to **delete** a field that is required in any possible `T` should produce the error `The operand of a 'delete' operator must be optional. (2790)`
2. **Smarter, more complex**: I can perform an update that is valid for some instantiations of `T`, but only after narrowing the type within the function body. Parametricity must also be preserved in the return type. 
    - Attempting to **overwrite** the `src` field would only be permissible if `x` has been refined to type `T extends Checkbox`, where `src` can be treated as a superfluous property under the inexact object subtyping model. The narrowed branch must return a value assignable to `T extends Checkbox`.
    - Attempting to **delete** the `userId` field would only be permissible if `x` has been refined to type `T extends { type: "guest" … }`, where `userId` is optional. The narrowed branch must return a value assignable to `T extends { type: "guest" … }`.

    Note that in both cases, we can narrow safely because `T` is constrained by a disjoint union of objects, where the standard depth and width subtyping rules apply. That is, while `T` could be instantiated with a different object like
    ```typescript
    { 
      type: "member", 
      name: "Nemo" | "Dory",
      userId: string,
      randomOtherField: boolean
    }
   ```
   this `T` must still contain all fields declared in the base `member` type, with types assignable to the corresponding fields.

There is also a third option for immutable updates, as in the first example:

3. **Allow updates, but refine the resulting type so it is not assignable to `T`**: TypeScript could allow me to create the value `{ ...x, src: 1 }`, but prevent me from returning this value as a `T`. 
    
    Currently, TypeScript considers `T & { src: number }` assignable to `T`. But if `T extends Image | Checkbox`, then the resulting intersection
    ```ts
    { type: "image", src: string & number }
    | { type: "checkbox", checked: boolean }
    ```
    cannot satisfy the original constraint, and therefore cannot be returned by the function.

For what it's worth, Flow uses Option 3 for the case of [spreading updates into a new object](https://flow.org/try/#1N4Igxg9gdgZglgcxALlAIwIZoKYBsD6uEEAztvhgE6UYCe+JADpdhgCYowa5kA0I2KAFcAtiRQAXSkOz9sADwxgJ+NPTbYuQ3BMnTZA+Y2yU4IwRO4A6SFBIrGVDGM7c+h46fNRLuKxJIGWh8MeT0ZfhYlCStpHzNsFBAMIQkIEQwJODAQfiEyfBE4eWw2fDgofDBMsAALfAA3KjgsXGxxZC4eAw0G-GhcWn9aY3wWZldu-g1mbGqJUoBaCRHEzrcDEgBrbAk62kXhXFxJ923d-cPRHEpTgyEoMDaqZdW7vKgoOfaSKgOKpqmDA+d4gB5fMA-P6LCCMLLQbiLOoYCqgh6-GDYRYIXYLSgkRZkCR4jpddwPfJLZjpOBkO4AX34kA0SQA9KyAAQAKh5vL5-IFgqFwpFPIAOlB2RyAKKKESMNrIDkAVRIFQQHIgaAAVnMJBymFE2BL2SbOQBxOANQQcjAcwEtNBtDkAA3kLo5AHdaqRsByVsYObT-bU-TivqYwP7Vvbmq0-S6ACou3hmr1+6pQA2zdhBqBNXBwNgcoSMNiZdp5tKu91e2pwZ2zMiUBrqoMBaOBpMuqxp0X9geDwUSiUBv0ASQyOI5AF4OcAJRyl53sEqxSAzBgcevU1nlyRKGAlfZTFAELv6SOfDGAMKhsBbNAQeSz+eL5djtfge+P58799LnUcw7GwSpPhAzxQBeV4wA8yhwNAJZ2BgmIADyJhyCgLFAbAkByk5bn6AA+HJ3sBT7yAAfAAFPISqJgAlPRb57oB0D2BytCvsAHJWHx8i8Aah5KgAjBy9IclKN5EGgNx4SwACOQhwCwxbwHgbAAIQARyLASEIlBZrQEqXlAEq2Bxm7TnO6Iodg1E8Z+HLrlZ2A7kJR7OSAtRUBBbBWAgcAwOu4kMVeADc1GuUqBE4gxkmco4JBkCQEqRa5VgHp5J7qvFUpJSl5nsRB2BWEQCBRVOpVZTYPmUAAghI1EAAwMXlnKypCcIIVASoZTVyINfqwZQBA+p2rBjzwlAuQgNa+I9UkDTNVYABMACs61WM1ID0kAA), and Option 1 for the case of [mutating the original argument](https://flow.org/try/#1N4Igxg9gdgZglgcxALlAIwIZoKYBsD6uEEAztvhgE6UYCe+JADpdhgCYowa5kA0I2KAFcAtiRQAXSkOz9sADwxgJ+NPTbYuQ3BMnTZA+Y2yU4IwRO4A6SFBIrGVDGM7c+h46fNRLuKxJIGWh8MeT0ZfhYlCStpHzNsFBAMIQkIEQwJODAQfiEyfBE4eWw2fDgofDBMsAALfAA3KjgsXGxxZC4eAw0G-GhcWn9aY3wWZldu-g1mbGqJUoBaCRHEzrcDEgBrbAk62kXhXFxJ923d-cPRHEpTgyEoMDaqZdW7vKgoOfaSKgOKpqmDA+d4gB5fMA-P6LCCMLLQbiLOoYCqgh6-GDYRYIXYLSgkRZkCR4jpddwPfJLZjpOBkO4AX34kA0SQA9KyAAQAKh5vL5-IFgqFwpFPIAOlB2RyAKKKESMNrIDkAVRIFQQHIABho2gtNRL2QbOQBxOANQQcjAcwEtNBtLXyTUcgDutVI2A5K2MHNpntqHpxX1MYE9q2tzVaHs1ABVNbwjS6PdUoBzPDAIJQRD6oE1cHA2Bydbs4NASBzoA6na64PbZmRKA11T6AqHvTHNVYE6Luz3e4KJRKvR6AIKpN2UDkAXg5Eo5c45AB8OcBW9glWKQOYRDcN7wOVBnGuOfZTFAEHvKZQAJJsJUnpv02fzpcrofrkAIGT2Xf7w93qTqhe9Y3gA-P+p4ao+UASjADzKCWKZFgsqomDeAA80YcgoCxQGwZajhI44AHwABTyEq0YAJQUcuT5zkhHryFYl43hyUoAErYAAjkIcAsAW6YTluNwkAAhHRHIsBIQiUCm8gSlBEq2PYHLCSYU6FnguzYCh15sCRwASW+M6btg24mLuEkHuY74AMp5iIlkpvOLG3iZYAAMxsAAnAAHDAHkAEyLIFWBoIsAAsGAAKzRYsGARd5ADsixsGgvlJdFGgwDA0XYNgTn0pRA5QAA3CRamUEqBHjpRbGco4JBkCQErlZVzHAW595nnVUqNc1SmlhAbRWEQCAVWZNwdahbA2LUVDDhIJEAAyUb1nKypCcIIUq7WuT6ZYPNlFSlLkIDmviCFJA0y1WIFsVWMtID0kAA).

**Comment by jcalz** [comment]
Note: not a TS team member.

Partially a duplicate of #50559, ultimately a likely duplicate of #9825 or similar (i.e., TS has many known holes in its type system that are present because plugging them makes things worse overall.)

**Comment by sliminality** [comment]
> `{ ...x, src: 1 }` is typed the way it is because that's the _closest_ thing to a correct representation currently available

While I agree that a type-level spread operator would be _optimal_, I don't think it's _necessary_ here. Even if we model `{ ...x, src: 1 }` as an intersection, ideally we could conclude that
```ts
(T extends Image | Checkbox) & { src: number }
```
ends up with `src` typed as `string & number`, which already reduces to `never` [when using concrete types](https://www.typescriptlang.org/play/?ts=5.7.2#code/C4TwDgpgBAkgtgQwObQLxQN4Cgq6qSALigCIBLRFEgGhzwGcAnAY2PuEbIDslaBfLFgLQAwgAsIzANYAjAPYAPKOmx584CMRLMJ0+Qpp1cOyVIgATYvLkAbCAi79B5yTYSNozOV3ZQFxeGRoAB8ocVN9QQAKDCgAOgSFaigmVigARig+FIRgMnoAMzIIeigowJQoUPC9RQBKKAAyTBSWYi4AVzgZCEYsuqA). Subtyping `T` can only make `src` narrower, and `"neat" & number` is still `never`.

> since `T` could be further subtyped, e.g. `T` could be
> 
> ```ts
> type MySpecialImage = {
>     type: "image",
>     src: "neat",
> }
> ```
>
> so it's not enough to validate that `src` is `string`

I agree it's not _enough_ to validate that `src` is `string`. I still think it would be a lot better than nothing (i.e. would help to catch errors in practice), and strictly sounder (i.e. consistent with the hypothetical perfect implementation, rather than a one-off hack that adds tech debt). To me this fits the TypeScript ethos of incremental improvements + not letting the perfect be the enemy of the good.

Fair enough on `delete`.
